### PR TITLE
refactor(Database): add method `remove` for `Database` & refactor the…

### LINKED ISSUE #33: insertHook 与 deleteHook 设计是否需要保留
`InsertHook` 最初的设计目的是为了在插入含 Virtual 类型字段的 table 的时候分离 Virtual 数据并插入数据表：

```ts
TaskSchemaDef: {
  _id: {
    type: RDBType.string,
    primary: true
  },
  project: {
    type: Association.oneToMany,
    virtual: {
      name: 'Project',
      where: projectTable => ({
        _projectId: projectTable._id
      })
    }
  }
}
```
但在实际应用场景中，并没有业务需要调用 insertHook

而对于 `deleteHook`，当初设计它是为了在某些数据删除的时候，同时删除与它关联的数据，比如`删除任务的时候同时删除它的子任务和关联`。但在实际业务中 hook 的定义非常难写，并且很难保持它的`原子性`。

所以根据以下的理由:

1. InsertHook 应用场景非常有限
2. DeleteHook 还没有一个非常合理的设计
3. 只有 insert 与 delete hook 对于 ReactiveDB 来说定义是不完整的
4. 很难保证这些 hook 的原子性

我们需要重新设计 Hook 吗，还是对外不暴露这些接口？

### PR DESCRIPTION
**Pull request by Saviio**
**BREAKING CHANGE:**
 - `Hook` mechanism is deprecated now [FIX #33]

this commit including:

[Feature]:
 - method: `remove`
 - method: `dump` 
 - method: `load`

[Refactor]
  - unified the returnValue for `CUD / UPSERT / REMOVE` method
  - deprecate `Hook` mechanism
  - rewrite part of test case for `Database`
  - improve the code-style

[Other]
- adjust the project structure
- adjust the output of test case

[Next]
  - rewrite comment
  - add a possible optimization for query which contain skip and limit (using parameterized query)
  - add `JoinMode` mechanism for more flexible self-join (i.e. Task -> Task)
  - inspector & middleware (maybe)

### CODE REVIEW

**Comment by Saviio** [comment]
还有些编译错误，等我明天rebase完一起修了.

File: src/utils/clone.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { forEach } from './for-each'
+
+export const clone = <T>(origin: T, target: T = null): T | null => {
```
Comment (Saviio) [inline src/utils/clone.ts]: 万恶的windows [doge]

File: src/exception/Exception.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+export interface ReactiveDBException extends Error { }
```
Comment (Saviio) [inline src/exception/Exception.ts]: 这个我重构的时候看了，主要是我对兼容性还有一点concern，所以暂时就没先实现。 「日本人」的说法也是有点道理的。

File: src/utils/clone.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  if (origin === null || origin === undefined) {
```
Comment (Saviio) [inline src/utils/clone.ts]: 主要是 `0` 也是个 `falsely` 的值，虽然 0 作为number直接return origin也行为上是对的，但我总觉得和这段代码想表达的语义有一点出入，所以通常我会用 ` == null || == undefined `

Comment (Brooooooklyn) [inline src/utils/clone.ts]: 0 == null， 0 == undefined 都是 false

**Pull request merged**